<a href="https://colab.research.google.com/github/feyza-inl/Su-Alti-Goruntu-Iyilestirme/blob/main/REAL_ESRGAN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install basicsr==1.4.2 --no-deps
!pip install realesrgan==0.3.0

# basicsr hatasını düzelt
dosya = '/usr/local/lib/python3.12/dist-packages/basicsr/data/degradations.py'
with open(dosya, 'r') as f:
    icerik = f.read()
icerik = icerik.replace(
    'from torchvision.transforms.functional_tensor import rgb_to_grayscale',
    'from torchvision.transforms.functional import rgb_to_grayscale'
)
with open(dosya, 'w') as f:
    f.write(icerik)
print('Hazır!')

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 172.5/172.5 kB 6.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for basicsr: filename=basicsr-1.4.2-py3-none-any.whl size=214819 sha256=a582c745812879c4ab8ef376f15a8747bdf56324413bfeed8bc63013622ab20b
  Stored in directory: /root/.cache/pip/wheels/9a/e3/e4/58f29bfabb622dd40b6d9839318ce5bf092062b81ca3aa19ea
Successfully built basicsr
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.8/46.8 kB 4.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 178.0/178.0 kB 8.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 59.6/59.6 kB 7.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 52.2/52.2 kB 6.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 299.4/299.4 kB 21.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.5/5.5 MB 86.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 256.2/256.2 kB 29.1 MB/s eta 0:

In [ ]:
!wget https://github.com/xinntao/Real-ESRGAN/releases/download/v0.1.0/RealESRGAN_x4plus.pth -P /content/weights/

import torch
from realesrgan import RealESRGANer
from basicsr.archs.rrdbnet_arch import RRDBNet

model = RRDBNet(num_in_ch=3, num_out_ch=3, num_feat=64, num_block=23, num_grow_ch=32, scale=4)
upsampler = RealESRGANer(
    scale=4,
    model_path='/content/weights/RealESRGAN_x4plus.pth',
    model=model,
    tile=400,
    tile_pad=10,
    pre_pad=0,
    half=True
)
print('Model hazır!')

--2026-03-04 11:21:11--  https://github.com/xinntao/Real-ESRGAN/releases/download/v0.1.0/RealESRGAN_x4plus.pth
Resolving github.com (github.com)... 140.82.113.4
Connecting to github.com (github.com)|140.82.113.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://release-assets.githubusercontent.com/github-production-release-asset/387326890/08f0e941-ebb7-48f0-9d6a-73e87b710e7e?sp=r&sv=2018-11-09&sr=b&spr=https&se=2026-03-04T11%3A58%3A51Z&rscd=attachment%3B+filename%3DRealESRGAN_x4plus.pth&rsct=application%2Foctet-stream&skoid=96c2d410-5711-43a1-aedd-ab1947aa7ab0&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skt=2026-03-04T10%3A58%3A00Z&ske=2026-03-04T11%3A58%3A51Z&sks=b&skv=2018-11-09&sig=jl4j7HYOicLgMcbdWahAuJtgFKtpc9%2BdRwRfOF9%2BQxA%3D&jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmVsZWFzZS1hc3NldHMuZ2l0aHVidXNlcmNvbnRlbnQuY29tIiwia2V5Ijoia2V5MSIsImV4cCI6MTc3MjYyNTA3MSwibmJmIjoxNzcyNjIzMjcxLCJwYXRoIjoicmVsZWFzZWFzc2V0cH

In [ ]:
import cv2
import os

raw_klasor = '/content/drive/MyDrive/raw-890'
iyilestirilmis_klasor = '/content/drive/MyDrive/iyilestirilmis-890'
os.makedirs(iyilestirilmis_klasor, exist_ok=True)

dosyalar = os.listdir(raw_klasor)
toplam = len(dosyalar)

for i, dosya in enumerate(dosyalar):
    raw_yol = os.path.join(raw_klasor, dosya)
    cikti_yol = os.path.join(iyilestirilmis_klasor, dosya)

    # Zaten işlendiyse atla
    if os.path.exists(cikti_yol):
        continue

    img = cv2.imread(raw_yol, cv2.IMREAD_COLOR)
    if img is None:
        continue

    output, _ = upsampler.enhance(img, outscale=2)
    cv2.imwrite(cikti_yol, output)
    print(f'{i+1}/{toplam} → {dosya}')

print('Tüm görüntüler iyileştirildi!')

	Tile 1/6
	Tile 2/6
	Tile 3/6
	Tile 4/6
	Tile 5/6
	Tile 6/6
1/890 → 186_img_.png
	Tile 1/6
	Tile 2/6
	Tile 3/6
	Tile 4/6
	Tile 5/6
	Tile 6/6
2/890 → 160_img_.png
	Tile 1/8
	Tile 2/8
	Tile 3/8
	Tile 4/8
	Tile 5/8
	Tile 6/8
	Tile 7/8
	Tile 8/8
3/890 → 15704.png
	Tile 1/4
	Tile 2/4
	Tile 3/4
	Tile 4/4
4/890 → 192_img_.png
	Tile 1/4
	Tile 2/4
	Tile 3/4
	Tile 4/4
5/890 → 157_img_.png
	Tile 1/8
	Tile 2/8
	Tile 3/8
	Tile 4/8
	Tile 5/8
	Tile 6/8
	Tile 7/8
	Tile 8/8
6/890 → 15544.png
	Tile 1/6
	Tile 2/6
	Tile 3/6
	Tile 4/6
	Tile 5/6
	Tile 6/6
7/890 → 19_img_.png
	Tile 1/4
	Tile 2/4
	Tile 3/4
	Tile 4/4
8/890 → 162_img_.png
	Tile 1/4
	Tile 2/4
	Tile 3/4
	Tile 4/4
9/890 → 159_img_.png
	Tile 1/4
	Tile 2/4
	Tile 3/4
	Tile 4/4
10/890 → 155_img_.png
	Tile 1/4
	Tile 2/4
	Tile 3/4
	Tile 4/4
11/890 → 194_img_.png
	Tile 1/4
	Tile 2/4
	Tile 3/4
	Tile 4/4
12/890 → 196_img_.png
	Tile 1/8
	Tile 2/8
	Tile 3/8
	Tile 4/8
	Tile 5/8
	Tile 6/8
	Tile 7/8
	Tile 8/8
13/890 → 20.png
	Tile 1/2
	Tile 2/2
14/890 → 200_img

KeyboardInterrupt: 

In [ ]:
import cv2
import numpy as np
import os
from skimage.metrics import peak_signal_noise_ratio as psnr
from skimage.metrics import structural_similarity as ssim

ref_klasor = '/content/drive/MyDrive/reference-890-extracted/reference-890'
iyilestirilmis_klasor = '/content/drive/MyDrive/iyilestirilmis-890'
raw_klasor = '/content/drive/MyDrive/raw-890'

psnr_raw_listesi = []
ssim_raw_listesi = []
psnr_iyi_listesi = []
ssim_iyi_listesi = []

for dosya in os.listdir(raw_klasor):
    raw_yol = os.path.join(raw_klasor, dosya)
    ref_yol = os.path.join(ref_klasor, dosya)
    iyi_yol = os.path.join(iyilestirilmis_klasor, dosya)

    if not os.path.exists(ref_yol) or not os.path.exists(iyi_yol):
        continue

    raw = cv2.imread(raw_yol)
    ref = cv2.imread(ref_yol)
    iyi = cv2.imread(iyi_yol)

    # Hepsini RAW boyutuna getir
    hedef_boyut = (raw.shape[1], raw.shape[0])
    ref_r = cv2.resize(ref, hedef_boyut)
    iyi_r = cv2.resize(iyi, hedef_boyut)  # İyileştirilmiş de küçültülüyor

    psnr_raw_listesi.append(psnr(ref_r, raw))
    ssim_raw_listesi.append(ssim(ref_r, raw, channel_axis=2))
    psnr_iyi_listesi.append(psnr(ref_r, iyi_r))
    ssim_iyi_listesi.append(ssim(ref_r, iyi_r, channel_axis=2))

print('========== SONUÇLAR ==========')
print(f'HAM GÖRÜNTÜ     → PSNR: {np.mean(psnr_raw_listesi):.2f} dB | SSIM: {np.mean(ssim_raw_listesi):.4f}')
print(f'İYİLEŞTİRİLMİŞ  → PSNR: {np.mean(psnr_iyi_listesi):.2f} dB | SSIM: {np.mean(ssim_iyi_listesi):.4f}')
fark_psnr = np.mean(psnr_iyi_listesi) - np.mean(psnr_raw_listesi)
fark_ssim = np.mean(ssim_iyi_listesi) - np.mean(ssim_raw_listesi)
print(f'FARK            → PSNR: {fark_psnr:+.2f} dB | SSIM: {fark_ssim:+.4f}')

========== SONUÇLAR ==========
HAM GÖRÜNTÜ     → PSNR: 16.58 dB | SSIM: 0.7428
İYİLEŞTİRİLMİŞ  → PSNR: 16.29 dB | SSIM: 0.6579
FARK            → PSNR: -0.30 dB | SSIM: -0.0850


outscale=1 → "Büyütme, sadece iyileştir"
→ Model daha az değişiklik yapıyor
→ Görsel fark az

outscale=4 → "4 kat büyüt"
→ Model çok daha fazla piksel dolduruyor
→ Bu sırada çok daha fazla iyileştirme yapıyor
→ Görsel fark belirgin!

PSNR yorumu: 16.58 → 16.29, çok küçük bir düşüş. Piksel seviyesinde model referanstan biraz uzaklaşmış ama fark minimal.SSIM yorumu: 0.7428 → 0.6579, bu daha ciddi. Yapısal benzerlik 0.085 düştü. Model görüntünün dokusunu, kenarlarını değiştirmiş — referansın beklediği yapıdan uzaklaşmış.Genel: Model görüntüyü işliyor ama sualtı renk bozulmasını düzeltmek yerine genel keskinleştirme yapıyor.

In [ ]:
import cv2
import os

raw_klasor = '/content/drive/MyDrive/raw-890'
iyi_klasor = '/content/drive/MyDrive/iyilestirilmis-890'

dosyalar = [f for f in os.listdir(iyi_klasor)][:5]

for dosya in dosyalar:
    raw = cv2.imread(os.path.join(raw_klasor, dosya))
    iyi = cv2.imread(os.path.join(iyi_klasor, dosya))

    if raw is None or iyi is None:
        continue

    kat = iyi.shape[0] / raw.shape[0]
    print(f'{dosya}:')
    print(f'  RAW: {raw.shape[1]}x{raw.shape[0]}')
    print(f'  İYİ: {iyi.shape[1]}x{iyi.shape[0]}')
    print(f'  Büyütme: {kat:.1f}x')
    print()

186_img_.png:
  RAW boyut: (771, 1200, 3)
  REF boyut: (771, 1200, 3)
  İYİ boyut: (1542, 2400, 3)
---
160_img_.png:
  RAW boyut: (1150, 800, 3)
  REF boyut: (1150, 800, 3)
  İYİ boyut: (2300, 1600, 3)
---
15704.png:
  RAW boyut: (720, 1280, 3)
  REF boyut: (720, 1280, 3)
  İYİ boyut: (1440, 2560, 3)
---
192_img_.png:
  RAW boyut: (480, 640, 3)
  REF boyut: (480, 640, 3)
  İYİ boyut: (960, 1280, 3)
---
157_img_.png:
  RAW boyut: (480, 640, 3)
  REF boyut: (480, 640, 3)
  İYİ boyut: (960, 1280, 3)
---


In [ ]:
import os

raw_klasor = '/content/drive/MyDrive/raw-890'
iyi_klasor = '/content/drive/MyDrive/iyilestirilmis-890'
ref_klasor = '/content/drive/MyDrive/reference-890-extracted/reference-890'

raw_dosyalar = set(os.listdir(raw_klasor))
iyi_dosyalar = set(os.listdir(iyi_klasor))
ref_dosyalar = set(os.listdir(ref_klasor))

# Üçünde de olan dosyalar
ortak = raw_dosyalar & iyi_dosyalar & ref_dosyalar
print(f'Üçünde de olan görüntü sayısı: {len(ortak)}')
print('Örnek:', list(ortak)[:5])

Üçünde de olan görüntü sayısı: 198
Örnek: ['517_img_.png', '15426.png', '206_img_.png', '19_img_.png', '138_img_.png']
